File: apps/spell-circle-canvas/src/common/compose/reference/examples/image_element.py
"""image — a picture as a leaf, and the three ways it meets its box.

The picture is built here rather than loaded, so the example depends on
no asset: an eight-by-four checker, sampled nearest, whose proportions
differ from the cells it is shown in.

The Python twin of image_element.cpp. A reference example: it is
rendered with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose, image, material, skia
from sigil.sketch import SketchContext, sketch

GROUND = "#14181d"
CELL = "#1c232a"
ASH = "#8ea0ad"
LIGHT = (0x6F, 0xB3, 0xA6, 0xFF)
DARK = (0x25, 0x30, 0x3A, 0xFF)


def checker() -> skia.Image:
    """Eight by four pixels of checker, twice as wide as it is tall, so a
    square box is a box the fit has to do something about."""
    pixels = bytearray()
    for row in range(4):
        for column in range(8):
            pixels.extend(LIGHT if (row + column) % 2 == 0 else DARK)
    return image.from_rgba(bytes(pixels), 8, 4)


def cell(caption: str, leaf: compose.Element) -> compose.Element:
    """The leaf is given a SQUARE box, not the cell's own. A leaf told to
    cover its parent has nothing left for the fit to decide; a box whose
    proportions disagree with the picture's is what makes the three
    answers different pictures."""
    return (
        compose.box(
            compose.box(leaf.width(120).height(120))
            .height(140)
            .width(compose.pct(100))
            .corners(8)
            .fill(CELL)
            .clip()
            .justify("center")
            .alignItems("center"),
            compose.text(caption, size=12, color=ASH),
        )
        .column()
        .gap(8)
        .basis(0)
        .grow(1)
        .alignItems("center")
    )


@sketch(size=(620, 260), background=GROUND, capture_at=0)
class ImageElement:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        picture = checker()
        # Nearest sampling on the container reaches every image leaf
        # under it, which is what a grid of pixels wants and a photograph
        # does not.
        return (
            compose.box(
                cell("Fit.Contain", compose.image(picture, material.Fit.Contain)),
                cell("Fit.Cover", compose.image(picture, material.Fit.Cover)),
                cell("Fit.Stretch", compose.image(picture, material.Fit.Stretch)),
            )
            .row()
            .gap(16)
            .padding(22)
            .sampling(skia.SamplingOptions(skia.FilterMode.Nearest))
        )

File: apps/spell-circle-canvas/src/common/compose/reference/examples/image_element.cpp
/** @file
 * image — a picture as a leaf, and the three ways it meets its box.
 *
 * The picture is drawn here rather than loaded, so the example depends on
 * no asset: a four-by-four checker, sampled nearest, whose proportions
 * differ from the cells it is shown in.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <include/core/SkCanvas.h>
#include <include/core/SkSurface.h>
#include <sigilcompose/core/Core.h>
#include <sigilsketch/canvas/Sketch.h>

namespace sketch = sigil::sketch;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {620, 260};
constexpr SkColor4f kGround = hexColor(0x14181d);
constexpr SkColor4f kCell = hexColor(0x1c232a);
constexpr SkColor4f kAsh = hexColor(0x8ea0ad);

/** Eight by four pixels of checker, so a square cell shows what each fit
 *  does with the proportions. */
sk_sp<SkImage> checker() {
  sk_sp<SkSurface> surface =
      SkSurfaces::Raster(SkImageInfo::MakeN32Premul(8, 4));
  SkCanvas& canvas = *surface->getCanvas();
  canvas.clear(hexColor(0x25303a).toSkColor());
  SkPaint paint;
  paint.setColor4f(hexColor(0x6fb3a6), nullptr);
  for (int y = 0; y < 4; ++y)
    for (int x = 0; x < 8; ++x)
      if ((x + y) % 2 == 0)
        canvas.drawRect(SkRect::MakeXYWH((float)x, (float)y, 1, 1), paint);
  return surface->makeImageSnapshot();
}

Element cell(const char* caption, Element leaf) {
  return box()
      .column()
      .gap(8)
      .basis(0)
      .grow(1)
      .alignItems(Align::Center)
      .children({box()
                     .height(140)
                     .width(pct(100))
                     .corners({8})
                     .fill(kCell)
                     .clip()
                     .children({std::move(leaf)}),
                 text(caption).font({.size = 12, .color = kAsh})});
}

}  // namespace

struct ImageElement {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    const sk_sp<SkImage> picture = checker();
    // Nearest sampling on the container reaches every image leaf under
    // it, which is what a grid of pixels wants and a photograph does not.
    return box()
        .row()
        .gap(16)
        .padding(22)
        .sampling(SkSamplingOptions(SkFilterMode::kNearest))
        .children({cell("Fit::Contain", image(picture, Fit::Contain).cover()),
                   cell("Fit::Cover", image(picture, Fit::Cover).cover()),
                   cell("Fit::Stretch", image(picture, Fit::Stretch).cover())});
  }
};

SIGIL_SKETCH(ImageElement, "Reference · Compose",
             "the image leaf under each of the three fits")
